src/frameobject.ts

```diff
@@ -141,6 +143,36 @@ class Py_FrameObject implements IPy_FrameObj {
             }
         }
     }
+
+    emptyStack() {
+        while (this.stack.length > 0) {
+            this.pop();
+        }
+    }
+    raise_exception_here(t: Thread, message: string, type: string): void {
+        var val: IPy_Object = (<any> builtins)[type];
+        t.exc = val;
+        this.push(val);
+        t.addToTraceback(message);
+        this.tryCatchException(t);
+    }
+
+    tryCatchException(t: Thread): boolean {
+        // Whenever an exception occurs, tries to find a handler and if it can't outputs the traceback 
+        // search in current frame
+        if (!find_exception_handler(this)) {
+            // Search for exception handler in previous frames
```

Comment:
The frame objects shouldn't know about / reference `Thread.stack` directly, or other fields of the object. The `Thread` class should conceal its implementation to outsiders, and reveal a clean abstraction through its methods (`asyncReturn`, `throwException`, etc).

While I understand crawling up the stack via `Thread.stack` *works*, it's not good software engineering.